### PR TITLE
[GSOC] Initialise the Arches Class

### PR DESCRIPTION
**Pull request by shishir-11**
#### Brief description of what is fixed or changed
Initialises the arches class and its method to help with the solver

#### Release Notes

* physics.continuum_mechanics
  * a parabolic arches structure is added

### CODE REVIEW

**Comment by Ishanned** [comment]
Hi @shishir-11, is the PR complete or are you going to add some more things?

**Comment by shishir-11** [comment]
I'll be adding a few more things , would inform once it's completed

**Comment by shishir-11** [comment]
@Ishanned @AdvaitPote I have done most of the stuff I need for the initial phase but I'm running into a bit of trouble while solving the parabolic equation , I need the value of 'a' to be negative, which I'm trying to implement.

**Comment by shishir-11** [comment]
image
Do you guys know why I'm failing this test, otherwise it is complete.

**Comment by Ishanned** [comment]
Hi @shishir-11, please refer this PR https://github.com/sympy/sympy/pull/25272/files 
You need to add docstrings for each method in the class and also expose the docstring in index.rst

File: sympy/physics/continuum_mechanics/arch.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            y0 = self._crown_y
+            parabola_eqn = a*(x-x0)**2 + y0 - y
+            eq1 = parabola_eqn.subs({x:self._left_support[0], y:self._left_support[1]})
+            solution = solve((eq1),(a))
+            parabola_eqn = solution[0]*(x-x0)**2 + y0
+            if(parabola_eqn.subs({x:self._right_support[0]}) != self._right_support[1]):
+                raise ValueError("provided coordinates of crown and supports are not consistent with parabolic arch")
+            self._shape_eqn = parabola_eqn
+
+        elif self._crown_x:
```
Comment (AdvaitPote) [inline sympy/physics/continuum_mechanics/arch.py:55]: If I try to create an `Arch` object with `crown_x=5` and `crown_y=5` and the supports `(0, 0)` and `(10, 0)`, I get no problem, but if I try to create an object with the same supports and only `crown_x=5`, it gives me a `ValueError`

```
>>> from sympy.physics.continuum_mechanics.arch import Arches
>>> a = Arches((0,0),(10,0),crown_x=5, crown_y=5)
>>> a.get_parabola_eqn
5 - (x - 5)**2/5
>>> a = Arches((0,0),(10,0),crown_x=5)
    .
    .
    .
    raise ValueError("parabolic arch cannot be constructed with the provided coordinates")
ValueError: parabolic arch cannot be constructed with the provided coordinates

```

File: sympy/physics/continuum_mechanics/arch.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        return self._right_support
+
+    @property
+    def get_reaction_force(self):
+        """
+        return the reaction forces generated
+        """
+        return self._reaction_force
+
+    def apply_load(self,order,label,x1,mag,x2=None,angle=None):
```
Comment (AdvaitPote) [inline sympy/physics/continuum_mechanics/arch.py:103]: Write somewhere what this `order` is in the docs of the class or this method.

File: sympy/physics/continuum_mechanics/arch.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        else:
+            self._loads[label] = mag
+            raise KeyError("no such load in the provided load type or load type does not exist")
+
+    def add_support(self,left_support,right_support):
+        support_types = ['roller','hinge']
+        if left_support not in support_types or right_support not in support_types:
+            raise ValueError("supports must only be roller or hinged")
+        self._supports['left'] = left_support
+        self._right_support['right'] = right_support
```
Comment (AdvaitPote) [inline sympy/physics/continuum_mechanics/arch.py:144]: Should be `self._supports['right']`

File: sympy/physics/continuum_mechanics/arch.py (hunk 1 above)
Comment (shishir-11) [reply sympy/physics/continuum_mechanics/arch.py:63]: That happens cause when we use crown_x = 5 , all values of crown_y are possible i.e. we will get multiple solutions to the parabolic equation.

Comment (AdvaitPote) [reply sympy/physics/continuum_mechanics/arch.py:63]: Go and change the Value error then. Will resolve the conversation then.

**Comment by AdvaitPote** [comment]
Try to clear the tests quickly. Will merge then so that you can open the solver PR.

**Comment by shishir-11** [comment]
I'm unable to pass tests for some reason , I have tried changing the brackets in the failing tests (values are same) how do I work around it ?

**Comment by AdvaitPote** [comment]
Update the mailmap and also are the tests added by you in `test_arch` working locally on your device.

**Comment by shishir-11** [comment]
> Update the mailmap and also are the tests added by you in `test_arch` working locally on your device.

The tests are not working locally , the values are same the only difference being the bracket placement which even after copy pasting is not working as it modifies it to some other format while testing.

**Comment by AdvaitPote** [comment]
The tests are passing for the first parabola equation but not the second.

**Comment by shishir-11** [comment]
I'm unable to fix the test locally , do you know what seems to be the problem ?

**Comment by AdvaitPote** [comment]
Even I am stuck. The parabola equation above is passing but not the second which is quite the same.

**Comment by AdvaitPote** [comment]
```
b = Arch((0,0),(10,2),crown_x=10)     
assert b.get_parabola_eqn == 2 - (x - 10)**2/50
```

Just add this example instead at the end

**Comment by Ishanned** [comment]
> > Update the mailmap and also are the tests added by you in `test_arch` working locally on your device.
> 
> The tests are not working locally , the values are same the only difference being the bracket placement which even after copy pasting is not working as it modifies it to some other format while testing. I'll make a seperate PR for the mailmap since it is not related to my PR or project.

@shishir-11 I suspect this has something to with solve(). Can you try using simplify after solve? I remember I once faced a similar issue and this did the trick for me

**Comment by shishir-11** [comment]
> > > Update the mailmap and also are the tests added by you in `test_arch` working locally on your device.
> > 
> > 
> > The tests are not working locally , the values are same the only difference being the bracket placement which even after copy pasting is not working as it modifies it to some other format while testing. I'll make a seperate PR for the mailmap since it is not related to my PR or project.
> 
> @shishir-11 I suspect this has something to with solve(). Can you try using simplify after solve? I remember I once faced a similar issue and this did the trick for me

yes , that worked. Thank you @Ishanned

**Comment by AdvaitPote** [comment]
LGTM!

**Comment by Ishanned** [comment]
> LGTM! @Ishanned

LGTM too... let's merge this one and move to the next step! Merging
@shishir-11 @AdvaitPote

**Comment by Ishanned** [comment]
One slight suggestion @shishir-11, please make your commit messages more descriptive. I know you can't make the message very long, but commit message like "arch test" can also be named something like "added simplify to get_parabola_eqn unit test in test_arch.py"
This will help reviewers easily find a specific commit in PRs with many commits, and will also be helpful for you in case you come across your old PRs (possibly nostalgically XD)

**Comment by shishir-11** [comment]
> One slight suggestion @shishir-11, please make your commit messages more descriptive. I know you can't make the message very long, but commit message like "arch test" can also be named something like "added simplify to get_parabola_eqn unit test in test_arch.py" This will help reviewers easily find a specific commit in PRs with many commits, and will also be helpful for you in case you come across your old PRs (possibly nostalgically XD)

I'll keep that in mind when creating new commit messages.

**Pull request merged**